### PR TITLE
Implement TX error checking and RMA retransmit

### PR DESCRIPTION
**Pull request by ztiffany**
Signed-off-by: Zach Tiffany ztiffany@cray.com

@hppritcha @sungeunchoi @jshimek @jswaro

### CODE REVIEW

**Comment by ztiffany** [comment]
I'm working on adding MSG retransmission now.  This is relatd to #337.

File: prov/gni/include/fi_ext_gni.h [hunk 1]
```diff
@@ -54,6 +54,8 @@ typedef enum dom_ops_val { GNI_MSG_RENDEZVOUS_THRESHOLD,
 			   GNI_MBOX_MSG_MAX_SIZE,
 			   GNI_RX_CQ_SIZE,
 			   GNI_TX_CQ_SIZE,
+			   GNI_MAX_RETRANSMITS,
```
Comment (sungeunchoi) [inline prov/gni/include/fi_ext_gni.h:57]: This should be added to the man page.  Also, will the MSG operations use the same value?

File: prov/gni/include/fi_ext_gni.h [hunk 2]
```diff
@@ -54,6 +54,8 @@ typedef enum dom_ops_val { GNI_MSG_RENDEZVOUS_THRESHOLD,
...
+			   GNI_ERR_INJECT_COUNT,
```
Comment (sungeunchoi) [inline prov/gni/include/fi_ext_gni.h:58]: Is GNI_ERR_INJECT_COUNT primarily for testing purposes?  If so, I'd leave it out of the man page.  We can discuss later if we want to compile it out for "production" builds.

File: prov/gni/src/gnix_msg.c [hunk 3]
```diff
@@ -287,7 +287,7 @@ static int __gnix_rndzv_req(void *arg)
  * GNI SMSG callbacks invoked upon completion of an SMSG message at the sender.
  ******************************************************************************/
```
Comment (sungeunchoi) [inline prov/gni/src/gnix_msg.c:288]: Please put in the comments above what the gni_return_t parameter is for.

File: prov/gni/src/gnix_ep.c [hunk 4]
```diff
@@ -321,11 +321,15 @@ ssize_t gnix_ep_msg_injectdata(struct fid_ep *ep, const void *buf, size_t len,
  * EP RMA API function implementations.
  ******************************************************************************/
 
+#define GNIX_RMA_READ_FLAGS_DEF		(FI_RMA | FI_READ)
+#define GNIX_RMA_WRITE_FLAGS_DEF	(FI_RMA | FI_WRITE)
+
```
Comment (sungeunchoi) [inline prov/gni/src/gnix_ep.c:326]: Are all these flags changes unrelated to retransmission?  Just something we should have been doing anyways?

File: prov/gni/src/gnix_rma.c [hunk 5]
```diff
@@ -79,17 +114,20 @@ static int __gnix_rma_send_completion(struct gnix_fid_ep *ep,
 				  "_gnix_cntr_inc() failed: %d\n", rc);
 	}
 
-	return rc;
+	return FI_SUCCESS;
```
Comment (sungeunchoi) [inline prov/gni/src/gnix_rma.c:117]: Is it documented somewhere that failed counter increment only is a warning?  If not, I think we should put it into our man page under CAVEATS.

File: prov/gni/src/gnix_ep.c (hunk 4 above)
Comment (ztiffany) [inline prov/gni/src/gnix_ep.c:326]: These are the default completion event flags for the two RMA types.  I'm not sure we were setting RMA flags in completions correctly before.  I now have better criterion tests for this.

File: prov/gni/src/gnix_rma.c (hunk 5 above)
Comment (ztiffany) [inline prov/gni/src/gnix_rma.c:117]: We could return an error here but there would be no way for it to propagate up to a user.  This function would be called inside the progress loop.

**Comment by hppritcha** [comment]
looks good.  Maybe default retry limit of 5 is a little low though?

**Pull request merged**